runtime/lua/vim/lsp/buf.lua

```diff
@@ -116,145 +116,92 @@ end
+--- vim.api.nvim_command[[autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting({save = true})]]
 --- </pre>
 ---
----@param options Table with valid `FormattingOptions` entries
----@param timeout_ms (number) Request timeout
----@see |vim.lsp.buf.formatting_seq_sync|
-function M.formatting_sync(options, timeout_ms)
-  local params = util.make_formatting_params(options)
-  local bufnr = vim.api.nvim_get_current_buf()
-  select_client('textDocument/formatting', function(client)
-    if client == nil then
-      return
-    end
-
-    local result, err = client.request_sync('textDocument/formatting', params, timeout_ms, bufnr)
-    if result and result.result then
-      util.apply_text_edits(result.result, bufnr)
-    elseif err then
-      vim.notify('vim.lsp.buf.formatting_sync: ' .. err, vim.log.levels.WARN)
+---@param opts|nil
+---        * formatting_options table with valid `FormattingOptions` entries
+---        * client_id number restrict formatting to client_id
+---        * timeout_ms number Request timeout
+---        * save bool save buffer after formatting
+---@see https://microsoft.github.io/language-server-protocol/specification#textDocument_formatting
+function M.formatting(opts)
+  opts = opts or {}
+  local client = vim.lsp.get_client_by_id(opts.client_id) or select_client("textDocument/formatting")
+  if client == nil then return end
+
+  local request_buf = vim.api.nvim_get_current_buf()
+  vim.api.nvim_buf_set_option(request_buf, 'modifiable', false)
+  vim.notify(string.format("Locking buffer (id: %d) while formatting.", request_buf))
+
+  local formatting_handler = function(_, result, ctx, _)
+    vim.api.nvim_buf_set_option(request_buf, 'modifiable', true)
+    if not result then return end
+    util.apply_text_edits(result, ctx.bufnr)
+    if opts.save then
+      vim.api.nvim_buf_call(request_buf, function() vim.api.nvim_command("noautocmd write!") end)
     end
-  end)
-end
-
---- Formats the current buffer by sequentially requesting formatting from attached clients.
----
---- Useful when multiple clients with formatting capability are attached.
----
---- Since it's synchronous, can be used for running on save, to make sure buffer is formatted
---- prior to being saved. {timeout_ms} is passed on to the |vim.lsp.client| `request_sync` method.
---- Example:
---- <pre>
---- vim.api.nvim_command[[autocmd BufWritePre <buffer> lua vim.lsp.buf.formatting_seq_sync()]]
---- </pre>
----
----@param options (optional, table) `FormattingOptions` entries
----@param timeout_ms (optional, number) Request timeout
----@param order (optional, table) List of client names. Formatting is requested from clients
----in the following order: first all clients that are not in the `order` list, then
----the remaining clients in the order as they occur in the `order` list.
-function M.formatting_seq_sync(options, timeout_ms, order)
-  local clients = vim.tbl_values(vim.lsp.buf_get_clients());
-  local bufnr = vim.api.nvim_get_current_buf()
+  end
```

Comment:
@mjlbach I currently have the following formatting function:

```lua
function M.format_buffer()
  if not util.has_lsp_clients() then
    return
  end

  local bufnr = api.nvim_win_get_buf(0)
  local windows = fn.win_findbuf(bufnr)
  local marks = {}

  -- Until https://github.com/neovim/neovim/issues/14645 is solved, I use this
  -- code to ensure the cursor position is properly restored after formatting a
  -- buffer. The approach used supports restoring cursors for different windows
  -- using the same buffer.
  for _, window in ipairs(windows) do
    local line, col = unpack(api.nvim_win_get_cursor(window))
    local ok, result = pcall(
      api.nvim_buf_set_extmark,
      bufnr,
      format_mark_ns,
      line - 1,
      col,
      {}
    )

    if ok then
      marks[window] = result
    end
  end

  lsp.buf.formatting_sync(nil, 5000)

  for _, window in ipairs(windows) do
    local mark = marks[window]

    local line, col = unpack(
      api.nvim_buf_get_extmark_by_id(bufnr, format_mark_ns, mark, {})
    )

    local max_line_index = api.nvim_buf_line_count(bufnr) - 1

    if line and col and line <= max_line_index then
      api.nvim_win_set_cursor(window, { line + 1, col })
    end
  end

  api.nvim_buf_clear_namespace(bufnr, format_mark_ns, 0, -1)
end
```

The fiddling with extmarks is necessary to properly restore the cursor position, see https://github.com/neovim/neovim/issues/14645 for details about this.

In the current stable setup this works, as we have a sync formatting function. In this new setup it won't work, as there's no way to run something after formatting finishes.

I could of course copy this entire function and call that instead of `vim.lsp.buf.formatting`, but I wonder if "run this function after formatting" is something that's worth supporting in this function here.

Reply:
I'd rather someone explicitly tackle https://github.com/neovim/neovim/issues/14645 than explicitly extend the API for the formatting helpers to work around a bug.